src/tools/w3dviewer/mainfrm.cpp

```diff
@@ -0,0 +1,1616 @@
+
+        if (dlg.DoModal() != IDOK) {
+            return;
+        }
+
+        Device = dlg.m_device;
+        BPP = dlg.m_bpp;
+
+        if (!view->Create()) {
+            return;
+        }
+    }
+}
+
+void CMainFrame::UpdateEmitterMenu()
+{
+    // do later
+}
+
+int CMainFrame::OnCreate(LPCREATESTRUCT lpCreateStruct)
+{
+    static UINT indicators[] = { 0, IDS_POLYS, IDS_PARTICLES, IDS_CAMERADIST, IDS_FRAMES, IDS_FRAMETIME, IDS_WWW };
+
+    theApp.m_pMainWnd = this;
+
+    if (CFrameWnd::OnCreate(lpCreateStruct) == -1 || !m_toolBar.Create(this) || !m_toolBar.LoadToolBar(IDR_MAINFRAME)
+        || !m_statusBar.Create(this) || !m_statusBar.SetIndicators(indicators, _countof(indicators))) {
+        return -1;
+    }
+
+    m_objectToolbar.Create("Object controls", this, ID_OBJECT_TOOLBAR);
+    m_objectToolbar.AddItem(IDB_XDIR, IDB_XDIRSEL, ID_CAMERA_ROTATEYONLY, TRUE);
+    m_objectToolbar.AddItem(IDB_YDIR, IDB_YDIRSEL, ID_CAMERA_ROTATEXONLY, TRUE);
+    m_objectToolbar.AddItem(IDB_ZDIR, IDB_ZDIRSEL, ID_CAMERA_ROTATEZONLY, TRUE);
+    m_objectToolbar.AddItem(IDB_ROTATEZ, IDB_ROTATEZSEL, ID_OBJECT_ROTATEZ, TRUE);
+
+    m_animationToolbar.Create("Animation controls", this, ID_ANIMATION_TOOLBAR);
+    m_animationToolbar.AddItem(IDB_PLAY, IDB_PLAYSEL, ID_ANIMATION_PLAY, TRUE);
+    m_animationToolbar.AddItem(IDB_STOP, IDB_STOPSEL, ID_ANIMATION_STOP, FALSE);
+    m_animationToolbar.AddItem(IDB_PAUSE, IDB_PAUSESEL, ID_ANIMATION_PAUSE, TRUE);
+    m_animationToolbar.AddItem(IDB_REVERSE, IDB_REVERSESEL, ID_ANIMATION_STEPBACK, FALSE);
+    m_animationToolbar.AddItem(IDB_FFWD, IDB_FFWDSEL, ID_ANIMATION_STEPFORWARD, FALSE);
+    m_animationToolbar.ShowWindow(SW_HIDE);
+
+    m_toolBar.SetBarStyle(m_toolBar.m_dwStyle | CBRS_TOOLTIPS | CBRS_FLYBY | CBRS_SIZE_DYNAMIC);
+    m_toolBar.ModifyStyle(0, TBSTYLE_FLAT);
+    m_toolBar.EnableDocking(CBRS_ALIGN_ANY);
+    EnableDocking(CBRS_ALIGN_ANY);
+    DockControlBar(&m_toolBar, nullptr);
+
+    RECT r;
+    GetWindowRect(&r);
+    FloatControlBar(&m_objectToolbar, CPoint(r.left + 10, r.bottom - 100), CBRS_ALIGN_LEFT);
+    FloatControlBar(&m_animationToolbar, CPoint(r.left + 210, r.bottom - 100), CBRS_ALIGN_LEFT);
+    ShowControlBar(&m_animationToolbar, FALSE, FALSE);
+
+    m_statusBar.SetPaneText(1, "");
+    m_statusBar.SetPaneText(2, "");
+    m_statusBar.SetPaneText(3, "");
+    m_statusBar.SetPaneText(4, "");
+    m_statusBar.SetPaneText(5, "");
+    m_statusBar.SetPaneText(6, "");
+
+    W3DAssetManager::Get_Instance()->Set_W3D_Load_On_Demand(true);
+    W3DAssetManager::Get_Instance()->Set_Activate_Fog_On_Load(true);
+    GetWindowRect(&m_windowRect);
+    HMENU menu = ::GetMenu(m_hWnd);
+    HMENU SubMenu = GetSubMenu(menu, 3);
+    m_subMenu = SubMenu;
+    MENUITEMINFO info;
+    memset(&info, 0, sizeof(info));
+    info.cbSize = sizeof(info);
+    info.fMask = MIIM_STRING;
+    char buffer[200];
+    info.dwTypeData = buffer;
+    info.cch = 100;
```

Comment:
I am very confused by this.

https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-menuiteminfoa

> cch is used only if the MIIM_TYPE flag is set in the fMask member

and

> The value of cch+1 is the size needed. Then allocate a buffer of this size, place the pointer to the buffer in dwTypeData, increment cch, and call GetMenuItemInfo once again

Reply:
fixed, the GetMenuItemInfo call wasn't even supposed to be there.